File: sdk/node/proof-profile.js
/* VELLA SDK — MIT License — Copyright (c) 2026 Vella Cognitive, LLC */
import proofV2 from "./proof-v2.cjs";

// Profiles are application-owned dependencies, never selected from a caller's proof.
export const V2_PROFILE = Object.freeze({
  id: "v2", receiptKind: "vella_execution_receipt_v1", recordFields: Object.freeze({}),
  digest: proofV2.digest, isDigest: value => typeof value === "string" && /^sha256:[0-9a-f]{64}$/.test(value),
  sign: proofV2.signV2, verify: proofV2.verifyV2, assertJson: proofV2.assertJson,
});

export function snapshotProofProfile(profile = V2_PROFILE) {
  if (typeof profile?.id !== "string" || !profile.id || typeof profile.receiptKind !== "string" ||
    !["digest", "isDigest", "sign", "verify", "assertJson"].every(name => typeof profile[name] === "function")) {
    throw new TypeError("invalid operator-owned proof profile");
  }
  proofV2.assertJson(profile.recordFields);
  return Object.freeze({ id: profile.id, receiptKind: profile.receiptKind,
    recordFields: Object.freeze(JSON.parse(JSON.stringify(profile.recordFields))),
    digest: profile.digest, isDigest: profile.isDigest, sign: profile.sign, verify: profile.verify, assertJson: profile.assertJson });
}

File: sdk/python/vella/proof_profile.py
"""Application-owned proof providers. Never selected by an untrusted envelope."""
from collections.abc import Mapping
from typing import Any, Protocol

from . import proof_v2


class ProofProfile(Protocol):
    @property
    def id(self) -> str: ...
    @property
    def record_fields(self) -> Mapping[str, Any]: ...
    def assert_json(self, value: object) -> None: ...
    def digest(self, value: Any) -> str: ...
    def sign(self, record: dict[str, Any], signing_key: object) -> dict[str, Any]: ...


class V2Profile:
    id = "v2"
    record_fields: Mapping[str, Any] = {}

    @staticmethod
    def assert_json(value: object) -> None:
        proof_v2.assert_json(value)

    @staticmethod
    def digest(value: Any) -> str:
        return proof_v2.digest(value)

    @staticmethod
    def sign(record: dict[str, Any], signing_key: object) -> dict[str, Any]:
        if not isinstance(signing_key, (str, bytes)):
            raise TypeError("v2 signing key must be PEM")
        return proof_v2.sign_v2(record, signing_key)
